### PR TITLE
Re-factoring download and extraction to enable hash based caching

### PR DESCRIPTION
**Pull request by fejese**
The `remote_file` helper supports sending headers with checksum (and timestamp) to prevent downloading files if they are present already and matching the one to be downloaded. By not moving the downloaded zip files before extracting we can let this optimisation kick in. With an initial chef run there's no difference. However if you want to update a server by running recipes, it accelerates the process if you don't have to download the files again, but only if they are matching. I disabled the modification date based caching, since it might occur that the download is aborted leaving a damaged file with newer timestamp in place.
If you feel that this change backward compatibility, I'm happy to change the default value for the attribute controlling this behaviour to keep downloading the file even if there's a match.

### CODE REVIEW

**Comment by anujbiyani** [comment]
partially addresses https://github.com/Lytro/aws_developer_tools/issues/13. the tools will still continually be extracted and installed when they don't need to be, but that can be dealt with separately.

**Comment by anujbiyani** [comment]
I think this will break backwards compatibility with Chef 10.x but that's okay. I can't find `use_conditional_get` in Chef 10.x's API.

File: definitions/cli_tools.rb [hunk 1]
```diff
@@ -5,11 +5,19 @@
 
   remote_file "/tmp/#{params[:name] + params[:extension]}" do
     source params[:source]
+    use_conditional_get !node['aws_developer_tools']['force_download?']
```
Comment (anujbiyani) [inline definitions/cli_tools.rb:8]: I think it's redundant to use `use_conditional_get` AND `use_etag` and `use_last_modified`. Just using the latter two should work, right?

```
use_etag !node['aws_developer_tools']['force_download?']
use_last_modified false
```

Comment (fejese) [inline definitions/cli_tools.rb:8]: Not exactly. At least as far as I understand you need to set `use_conditional_get` to `true` to enable the two other settings. As the documentation says for both `use_etag` and `use_last_modified`:

> To use this setting, `use_conditional_get` must also be set to `true`

So basically `use_conditional_get` enables or disables any caching and the other two flags enables specific caching methods.

That being said `use_conditional_get` is `true` by default so actually your proposed solution would do the same for now but I find it less clear why :)

**Comment by fejese** [comment]
Regarding #13 : I think this is as far as we can get with fixing it without too much overhead.

Since the process can fail at any point it might happen that the extraction fails. In that case some files could be missing or corrupt. I don't see an easy way to make sure that all the following are met:
- if a script is removed from the archive it needs to be removed from the system
- if a new script is added it needs to be added
- if a script is different it needs to be updated
- if a script exists and matches it needs to be left alone

It could be possible to create a flag after successful extraction and that can be checked and the time stamp of it could be compared to the downloaded file's but that's probably not worth it if there's no built in support for this kind of operation in chef.

File: definitions/cli_tools.rb (hunk 1 above)
Comment (anujbiyani) [inline definitions/cli_tools.rb:8]: Ah, I see. Looks fine then

**Comment by anujbiyani** [comment]
Thanks! The next release will have the major version bumped and I'll update the README to say we're now only compatible with Chef 11+

**Comment by fejese** [comment]
Cheers

**Pull request merged**